src/coreclr/src/jit/rangecheck.cpp

```diff
@@ -773,11 +773,46 @@ void RangeCheck::MergeAssertion(BasicBlock* block, GenTree* op, Range* pRange DE
 // Compute the range for a binary operation.
 Range RangeCheck::ComputeRangeForBinOp(BasicBlock* block, GenTreeOp* binop, bool monIncreasing DEBUGARG(int indent))
 {
-    assert(binop->OperIs(GT_ADD));
+    assert(binop->OperIs(GT_ADD, GT_AND, GT_RSH, GT_UMOD));
 
     GenTree* op1 = binop->gtGetOp1();
     GenTree* op2 = binop->gtGetOp2();
 
+    if (binop->OperIs(GT_AND, GT_RSH, GT_UMOD))
+    {
+        if (!op2->IsIntCnsFitsInI32())
+        {
+            // only cns is supported for op2 at the moment for &,%,>> operators
+            return Range(Limit::keUnknown);
+        }
+
+        int icon = -1;
+        if (binop->OperIs(GT_AND))
+        {
+            // x & cns -> [0..cns]
+            icon = static_cast<int>(op2->AsIntCon()->IconValue());
+        }
+        else if (binop->OperIs(GT_UMOD))
+        {
+            // x % cns -> [0..cns-1]
+            icon = static_cast<int>(op2->AsIntCon()->IconValue()) - 1;
+        }
+        else if (binop->OperIs(GT_RSH) && op1->OperIs(GT_AND) && op1->AsOp()->gtGetOp2()->IsIntCnsFitsInI32())
+        {
+            // (x & cns1) >> cns2 -> [0..cns1>>cns2]
+            icon = static_cast<int>(op1->AsOp()->gtGetOp2()->AsIntCon()->IconValue()) >>
+                   static_cast<int>(op2->AsIntCon()->IconValue());
```

Comment:
Shouldn't this only calculate the shift if `op2` is in the `[0..31]` range?  Otherwise, that's UB in C++.

Reply:
@lpereira I have a test:
```csharp
for (int i = -1000; i < 1000; i++)
{
    for (int j = -1000; j < 1000; j++)
    {
        for (int k = -1000; k < 1000; k++)
        {
            int actualIndex = (i & j) >> k;
            int expectedLimit = j >> k;
            if (expectedLimit < 0)
                continue;

            if (actualIndex > expectedLimit)
            {
                throw new Exception("");
            }
        }
    }
}
```

Reply:
It's still undefined behavior to shift anything outside the range.  If it "works" with this particular compiler, with these particular optimization options, there's nothing stopping it from having the same results in the future or with a different compiler.

Reply:
@lpereira yeah you're right